xbmc/addons/AddonRepos.cpp

```diff
@@ -229,3 +237,39 @@ bool CAddonRepos::FindAddonAndCheckForUpdate(
 
   return false;
 }
+
+bool CAddonRepos::GetLatestVersionByMap(const std::string& addonId,
+                                        const std::map<std::string, std::shared_ptr<IAddon>>& map,
+                                        std::shared_ptr<IAddon>& result) const
+{
+  const auto& remote = map.find(addonId);
+  if (remote != map.end()) // is addon in the desired map?
+  {
+    result = remote->second;
+    return true;
+  }
+
+  return false;
+}
+
+bool CAddonRepos::GetLatestAddonVersion(const std::string& addonId,
+                                        std::shared_ptr<IAddon>& result) const
+{
+  const AddonRepoUpdateMode updateMode =
+      CAddonSystemSettings::GetInstance().GetAddonRepoUpdateMode();
+
+  if (updateMode == AddonRepoUpdateMode::OFFICIAL_ONLY)
+  {
+    return GetLatestVersionByMap(addonId, m_latestOfficialVersions, result);
+  }
+  else if (updateMode == AddonRepoUpdateMode::ANY_REPOSITORY)
+  {
+    if (!GetLatestVersionByMap(addonId, m_latestOfficialVersions, result))
+    {
+      if (!GetLatestVersionByMap(addonId, m_latestPrivateVersions, result))
+        return false;
+    }
+  }
+
+  return true;
+}
```

Comment:
Is this correct? `FindInstallableById()` will use this function to find an addon that has not been installed yet. But if we don't know which repo this addon is in (maybe it's only in a private repo) we would never install it. Probably need to check if it's in the official repos first like we do in other places.

So like: 

```cpp
bool hasOfficialVersion =
      GetLatestVersionByMap(addonId, m_latestOfficialVersions, result);

if (hasOfficialVersion)
{
  if (updateMode == AddonRepoUpdateMode::ANY_REPOSITORY)
  {
    std::shared_ptr<IAddon> thirdPartyAddon;

    // only use this version if it's higher than the official one
    if (GetLatestVersionByMap(addonId, m_latestPrivateVersions, thirdPartyAddon))
    {
      if (thirdPartyAddon->Version() > result->Version())
        result = thirdPartyAddon;
    }
  }
}
else
{
  if (!GetLatestVersionByMap(addonId, m_latestPrivateVersions, result))
    return false;
}
```

I think maybe this function call needs a new name. Maybe `GetLatestAddonVersionFromAllRepos()`. Then it's very clear this covers uninstalled add-ons too.

Reply:
that's correct. function should be renamed and logic changed

Reply:
renamed and logic changed